e2e/nomostest/registryproviders/local.go

```diff
@@ -0,0 +1,60 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package registryproviders
+
+import (
+	"fmt"
+
+	"kpt.dev/configsync/e2e"
+	"kpt.dev/configsync/e2e/nomostest/portforwarder"
+)
+
+// LocalProvider refers to the test registry-server running on the same test cluster.
+type LocalProvider struct {
+	// PortForwarder is a port forwarder to the in-cluster registry server.
+	// This is used to communicate from the tests to the in-cluster registry server.
+	PortForwarder *portforwarder.PortForwarder
+}
+
+// Type returns the provider type.
+func (l *LocalProvider) Type() string {
+	return e2e.Local
+}
+
+// OCIPushURL returns a URL for pushing images to the remote registry.
+// name refers to the repo name in the format of <NAMESPACE>/<NAME> of RootSync|RepoSync.
+func (l *LocalProvider) OCIPushURL(name string) (string, error) {
```

Comment:
OCI prefix on these methods seems a little redundant. Is it informative?

Reply:
The intention is to add `HelmPush` type commands as well with a subsequent PR, which will have different behavior

Reply:
Helm and OCI methods on the same Registry API? They seem like they should probably have different interfaces. Do they need to be the same?

Reply:
They don't need to be the same. My expectation is that this will be simpler and more performant. Keep in mind the `RegistryProvider` is mainly just returning the URLs for another interface to consume. In this case the LocalProvider returns an `OCIPushURL` that is consumed by `OCIImage` to push images with `crane`.

In the `helm` scenario, the LocalProvider will return a `HelmPushURL` that will be consumed by a `Helm` (name TBD) struct to push helm packages.

Reply:
I guess I expected a LocalOCIRegistry and a LocalHelmRegistry, with OCIRegistry and HelmRegistry interfaces, so the methods wouldn't need the context in them.

Reply:
That way we also wouldn't need the SyncSource enum. We could just may the `spec.sourceFormat` as an enum that maps 1:1 with a SourceProvider.

Reply:
Okay I rewrote this to use an OCIProvider concept rather than `RegistryProvider`, such that we will add a `HelmProvider` variant once the helm-sync api is introduced.

The `SyncSource` enum is still needed because that has to do with the fact that there is a single registry deployment on the cluster that is used for multiple source types. i.e. helm-sync and oci-sync may use the same registry, and that registry serves from a single endpoint/certificate - so both helm-sync and oci-sync would need to use the same endpoint/certificate under the hood.

Reply:
What you have here is probably fine.

I think what I was imagining was simply that there would be an OCIRegistry and a HelmRegistry, and the HelmRegistry would wrap the OCIRegistry. Then the test setup would start both, and register cleanup to shut down both, but the OCIRegistry would de-dupe the calls.